airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +51,226 @@ public void checkpoint() {
+                    currentDestination.getParent().toFile().mkdirs();
+                    Files.copy(sources.get(i), destinations.get(i));
+                }
+            }
+        }
+
+        LOG.info("HSQLDB database backup complete.");
+        return destination;
+    }
+
+    /**
+     * Perform an in-place database upgrade from HSQLDB 1.x to 2.x.
+     */
+    private void performHsqldbDatabaseUpgrade() throws SQLException {
+        // This will upgrade HSQLDB from 1.8 to 2.5 on the first connection.
+        try (Connection conn = DriverManager.getConnection(settingsService.getDefaultJDBCUrl())) {
+            LOG.info("Database connection established. Current version is: {}.", conn.getMetaData().getDatabaseProductVersion());
+            // On upgrade, the official documentation recommends that we
+            // run 'SHUTDOWN SCRIPT' to compact all the database into a
+            // single SQL file.
+            //
+            // In practice, if we don't do that, we did not observe issues
+            // immediately but after the upgrade.
+            LOG.info("Shutting down database...");
+            try (Statement st = conn.createStatement()) {
+                st.execute("SHUTDOWN SCRIPT");
+            }
+        }
+
+        LOG.info("Database connection closed. Current version is: {}.", getHsqldbDatabaseVersion());
+    }
+
+    /**
+     * Checks if the database is correctly shut down
+     */
+    private boolean isHsqldbDatabaseCorrectlyShutdown() {
+
+        // The db/airsonic.data file may or may not exist (contains CACHED tables).
+        // The db/airsonic.backup file may or may not exist (not used during normal operation).
+
+        // The db/airsonic.properties file must exist (contains db properties).
+        if (!new File(settingsService.getDefaultJDBCPath() + ".properties").exists()) {
+            LOG.warn("The db/airsonic.properties file does not exist. The database may be corrupt.");
+            return false;
+        }
+        // The db/airsonic.script file must exist (contains db data from all tables except CACHED).
+        if (!new File(settingsService.getDefaultJDBCPath() + ".script").exists()) {
+            LOG.warn("The db/airsonic.script file does not exist. The database may be corrupt.");
+            return false;
+        }
+        // The db/airsonic.log file should not exist (contains db logs that are not supported in the upgrade).
+        if (new File(settingsService.getDefaultJDBCPath() + ".log").exists()) {
+            LOG.warn("The db/airsonic.log file exists. The database may not be shut down correctly.");
+            return false;
+        }
+        // The db/airsonic.lck file should not exist (lock file, indicates another running process or unclean shutdown).
+        if (new File(settingsService.getDefaultJDBCPath() + ".lck").exists()) {
+            LOG.warn("The db/airsonic.lck file exists. The database may not be shut down correctly.");
+            return false;
+        }
```

Comment:
I don't think any of these checks are needed. All they seem to be doing is logging about the state of the db, which I don't see how it is useful to us, or even a concern, considering there isn't anything we're doing about it. It's just slowing up startup time for the sake of a few warns.

Consider this:
 - DB is corrupt. we issue a warn, and then what? we upgrade the db
 - DB wasn't shutdown. we issue a warn and then what? we upgrade the db
 - DB is fine. we don't issue anything, and then what? we upgrade the db

Everything leads to the same outcome. I would consider them useful if they were altering the decision to upgrade (to follow the decision tree), but they're not. 

I think DB monitoring is a valuable tool, but a different issue. It's probably good to know your DB is corrupt or wasn't shutdown properly, but that's an issue to be taken care of regardless of whether an upgrade is happening (using monitoring tools).

Why does the user need to know the status of his db at the time of the upgrade and not otherwise?
If it was corrupt, and we weren't upgrading, would we have issued a warn? (we don't)

Reply:
The presence of a `.log` file at least should be checked and logged, because it will lose data during the upgrade process. I'll look at reducing these checks.

Reply:
Hmmm, first, I'm not convinced it'll lose data if there is a log file. What makes you say that? Compaction isn't a requirement for an upgrade, is it? Is there some documentation that says you'll lose data? I mostly upgrade without compaction, and I don't think I've ever lost data due to an upgrade. It's usually _helpful_ and therefore _recommended_ to have a compact database when upgrading, but why is there fear of data loss?

Second, this method is called twice (before and after the upgrade). Each time you're not affecting the course of the decision-making.

Okay, let's say there is a log file, and let's say there is data loss involved if there is one. So what? What are we going to do by checking for it?
It's not like we're going to be able to stop the upgrade from happening.
Either way this upgrade is happening _no matter what_, right?

So what does checking for a db log file accomplish? What is the purpose? Just to log it to a logger?
Logging does what? Are the users going to be able to recover the data from the log file that they wouldn't otherwise if we didn't have that check or didn't log it? That's certainly not true.

This seems to be a case of logging creep. I really don't think this entire method is needed. Log files or no log files, however or whatever state the db is in, it's getting upgraded, there is simply no choice. Checking for presence or absence of files, certainly in my opinion, doesn't accomplish anything. If the driver doesn't find certain files, it'll throw an exception itself. Us logging about it beforehand and then feeding it to the driver anyway just adds another redundant check.

Reply:
The whole point of this, in my opinion, was to help figuring out what happened during a borked upgrade. It's mostly for forensics, and will not help at all in recovering the data. I'll remove them then.